# Chapter 18

In [1]:
import pandas as pd
import statsmodels.api as sm

In [2]:
# Exercise 1
wageprc = pd.read_stata("./stata/WAGEPRC.DTA")
X = sm.add_constant(wageprc[["gwage", "gprice_1"]])
model = sm.OLS(wageprc.gprice, X, missing="drop").fit()
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                 gprice   R-squared:                       0.454
Model:                            OLS   Adj. R-squared:                  0.450
Method:                 Least Squares   F-statistic:                     116.8
Date:                Sun, 30 Oct 2022   Prob (F-statistic):           1.21e-37
Time:                        23:18:36   Log-Likelihood:                 1298.5
No. Observations:                 284   AIC:                            -2591.
Df Residuals:                     281   BIC:                            -2580.
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          0.0013      0.000      4.675      0.000       0.001       0.002
gwage          0.0814      0.031      2.620      0.009       0.020       0.143
gprice_1       0.6396      0.045     14.331      0.000       0.552       0.727
==============================================================================
Omnibus:                       51.043   Durbin-Watson:                   2.363
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              378.052
Skew:                           0.433   Prob(JB):                     8.07e-83
Kurtosis:                       8.586   Cond. No.                         304.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [3]:
model.params[1] / (1 - model.params[2])

0.2259784476690603

In [4]:
X = sm.add_constant(wageprc[["gwage", "gprice_1", "gwage_1"]])
model = sm.OLS(wageprc.gprice, X, missing="drop").fit()
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                 gprice   R-squared:                       0.460
Model:                            OLS   Adj. R-squared:                  0.454
Method:                 Least Squares   F-statistic:                     79.44
Date:                Sun, 30 Oct 2022   Prob (F-statistic):           3.30e-37
Time:                        23:18:36   Log-Likelihood:                 1300.1
No. Observations:                 284   AIC:                            -2592.
Df Residuals:                     280   BIC:                            -2578.
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          0.0011      0.000      3.613      0.000       0.000       0.002
gwage          0.0898      0.031      2.866      0.004       0.028       0.151
gprice_1       0.6186      0.046     13.422      0.000       0.528       0.709
gwage_1        0.0551      0.032      1.740      0.083      -0.007       0.117
==============================================================================
Omnibus:                       50.774   Durbin-Watson:                   2.316
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              347.067
Skew:                           0.463   Prob(JB):                     4.32e-76
Kurtosis:                       8.336   Cond. No.                         325.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [5]:
(model.params[1] + model.params[3]) / (1 - model.params[2])

0.3799599608144597

C1.i Estimated IP is 0.081 and estimated LRP is 0.226.

C1.ii The IP in the previous problem is 0.119 and the LRP is 1.172. Both are much higher than the geometric distributed lag.

C1.iii The estimated IP is 0.09 and the LRP is 0.380. These are closer to (but higher than) (i) and still well below the results from 5.11 (previous problem).

In [6]:
# Exercise 2
hseinv = pd.read_stata("./stata/HSEINV.DTA")
hseinv["ginvpc_1"] = hseinv["ginvpc"].shift(1)
hseinv["ginvpc_2"] = hseinv["ginvpc"].shift(2)
X = sm.add_constant(hseinv[["linvpc_1", "ginvpc_1", "ginvpc_2", "t"]])
model = sm.OLS(hseinv.ginvpc, X, missing="drop").fit()
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                 ginvpc   R-squared:                       0.437
Model:                            OLS   Adj. R-squared:                  0.371
Method:                 Least Squares   F-statistic:                     6.592
Date:                Sun, 30 Oct 2022   Prob (F-statistic):           0.000486
Time:                        23:18:36   Log-Likelihood:                 31.647
No. Observations:                  39   AIC:                            -53.29
Df Residuals:                      34   BIC:                            -44.98
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         -0.7864      0.170     -4.626      0.000      -1.132      -0.441
linvpc_1      -0.9558      0.198     -4.833      0.000      -1.358      -0.554
ginvpc_1       0.5317      0.162      3.291      0.002       0.203       0.860
ginvpc_2       0.2900      0.165      1.761      0.087      -0.045       0.625
t              0.0068      0.002      3.177      0.003       0.002       0.011
==============================================================================
Omnibus:                        0.717   Durbin-Watson:                   1.771
Prob(Omnibus):                  0.699   Jarque-Bera (JB):                0.330
Skew:                          -0.223   Prob(JB):                        0.848
Kurtosis:                       3.058   Cond. No.                         421.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [7]:
hseinv["gprice_1"] = hseinv["gprice"].shift(1)
hseinv["gprice_2"] = hseinv["gprice"].shift(2)
X = sm.add_constant(hseinv[["lprice_1", "gprice_1", "gprice_2", "t"]])
model = sm.OLS(hseinv.gprice, X, missing="drop").fit()
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                 gprice   R-squared:                       0.200
Model:                            OLS   Adj. R-squared:                  0.106
Method:                 Least Squares   F-statistic:                     2.131
Date:                Sun, 30 Oct 2022   Prob (F-statistic):             0.0983
Time:                        23:18:36   Log-Likelihood:                 105.65
No. Observations:                  39   AIC:                            -201.3
Df Residuals:                      34   BIC:                            -193.0
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         -0.0404      0.019     -2.071      0.046      -0.080      -0.001
lprice_1      -0.2216      0.092     -2.409      0.022      -0.409      -0.035
gprice_1       0.3276      0.155      2.111      0.042       0.012       0.643
gprice_2       0.1301      0.149      0.872      0.389      -0.173       0.433
t              0.0010      0.000      1.995      0.054   -1.82e-05       0.002
==============================================================================
Omnibus:                        1.867   Durbin-Watson:                   1.627
Prob(Omnibus):                  0.393   Jarque-Bera (JB):                0.914
Skew:                          -0.064   Prob(JB):                        0.633
Kurtosis:                       3.739   Cond. No.                     1.46e+03
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.46e+03. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

C2.i The t-statistic on linvpc_1 is -4.833 which is larger (in absolute value) than -3.41 from Table 18.3 and, in fact, the 1% critical value of -3.96. We can reject the null hypothesis that there is a unit root.

C2.ii Using the same approach the t-statistic is -2.409 which is less (in absolute value) than any critical value we would want to use. We do not have evidence to reject the null hypothesis that there is a unit root for price.

C2.iii The first test gives evidence that there is not a unit root, while the possibility is still open for log(price). It would not make sense to test for cointegration in this case.

In [8]:
# Exercise 3
volat = pd.read_stata("./stata/VOLAT.DTA")
X = sm.add_constant(volat[["pcip_1", "pcip_2", "pcip_3"]])
model = sm.OLS(volat["pcip"], X, missing="drop").fit()
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                   pcip   R-squared:                       0.166
Model:                            OLS   Adj. R-squared:                  0.161
Method:                 Least Squares   F-statistic:                     36.40
Date:                Sun, 30 Oct 2022   Prob (F-statistic):           1.81e-21
Time:                        23:18:37   Log-Likelihood:                -2167.7
No. Observations:                 554   AIC:                             4343.
Df Residuals:                     550   BIC:                             4361.
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          1.8042      0.548      3.292      0.001       0.728       2.881
pcip_1         0.3491      0.043      8.210      0.000       0.266       0.433
pcip_2         0.0708      0.045      1.575      0.116      -0.017       0.159
pcip_3         0.0674      0.043      1.584      0.114      -0.016       0.151
==============================================================================
Omnibus:                      184.426   Durbin-Watson:                   1.999
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             1740.263
Skew:                           1.179   Prob(JB):                         0.00
Kurtosis:                      11.356   Cond. No.                         19.4
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [9]:
volat["pcip_4"] = volat["pcip_3"].shift(1)
X = sm.add_constant(volat[["pcip_1", "pcip_2", "pcip_3", "pcip_4"]])
model = sm.OLS(volat["pcip"], X, missing="drop").fit()
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                   pcip   R-squared:                       0.166
Model:                            OLS   Adj. R-squared:                  0.160
Method:                 Least Squares   F-statistic:                     27.24
Date:                Sun, 30 Oct 2022   Prob (F-statistic):           1.22e-20
Time:                        23:18:37   Log-Likelihood:                -2164.2
No. Observations:                 553   AIC:                             4338.
Df Residuals:                     548   BIC:                             4360.
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          1.7873      0.555      3.221      0.001       0.697       2.877
pcip_1         0.3494      0.043      8.179      0.000       0.265       0.433
pcip_2         0.0702      0.045      1.556      0.120      -0.018       0.159
pcip_3         0.0658      0.045      1.457      0.146      -0.023       0.154
pcip_4         0.0043      0.043      0.101      0.920      -0.080       0.088
==============================================================================
Omnibus:                      184.693   Durbin-Watson:                   1.999
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             1741.966
Skew:                           1.184   Prob(JB):                         0.00
Kurtosis:                      11.366   Cond. No.                         21.1
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [10]:
X = sm.add_constant(volat[["pcip_1", "pcip_2", "pcip_3", "pcsp_1", "pcsp_2", "pcsp_3"]])
model = sm.OLS(volat["pcip"], X, missing="drop").fit()
model.f_test("pcsp_1 = pcsp_2 = pcsp_3 = 0")

<class 'statsmodels.stats.contrast.ContrastResults'>
<F test: F=5.371828372897602, p=0.001190404630174995, df_denom=547, df_num=3>

In [11]:
volat["ci3_3"] = volat["ci3_2"].shift(1)
X = sm.add_constant(volat[["pcip_1", "pcip_2", "pcip_3", "pcsp_1", "pcsp_2", "pcsp_3", "ci3_1", "ci3_2", "ci3_3"]])
model = sm.OLS(volat["pcip"], X, missing="drop").fit()
model.f_test("pcsp_1 = pcsp_2 = pcsp_3 = 0")

<class 'statsmodels.stats.contrast.ContrastResults'>
<F test: F=5.076975015753374, p=0.0017866714063804869, df_denom=544, df_num=3>

C3.i Results above. The 4th lag is significant at the 95% level (i.e. not at all).

C3.ii The joint significance test is significant at the 1% level. The null hypothesis is that all of the coefficients for pcsp are 0 which we reject. Granger causality is defined in 18.51 ($E(y_t|I_{t-1} \neq E(y_t|J_{t-1})$) and, since we have strong evidence that at least one of the coefficients for pcsp is not zero, we conclude that pscp Granger causes pcip.

C3.iii Adding the change in the three-month T-bill does shrink the F-statistic (from 5.37 to 5.08, p-value 0.002) but the joint hypothesis test is still significant even at the 1% level. We still conclude pcsp Granger causes pcip conditional on past $\delta i3$.